core/app/app.go

```diff
@@ -0,0 +1,213 @@
+type LocalEvent struct {
+	Node format.Node
+	Auth *thread.Auth
+}
+
+// LocalEventListener notifies about new locally generated events.
+type LocalEventListener struct {
+	listener *broadcast.Listener
+	c        chan *LocalEvent
+}
+
+// Channel returns an unbuffered channel to receive local events.
+func (l *LocalEventListener) Channel() <-chan *LocalEvent {
+	return l.c
+}
+
+// Discard indicates that no further events will be received
+// and ready for being garbage collected.
+func (l *LocalEventListener) Discard() {
+	l.listener.Discard()
+}
+
+// Net adds the ability to connect an app to a thread.
+type Net interface {
+	net.Net
+
+	// ConnectApp returns an app<->thread connector.
+	ConnectApp(App, thread.ID) (*Connector, error)
+}
+
+// Connector connects an app to a thread.
+type Connector struct {
+	app        App
+	net        Net
+	threadID   thread.ID
+	threadKey  thread.Key
+	logID      peer.ID
+	closeChan  chan struct{}
+	goRoutines sync.WaitGroup
+
+	lock   sync.Mutex
+	closed bool
+}
+
+// Connection receives new thread records, which are pumped to the app.
+type Connection func(context.Context, map[thread.ID]struct{}) (<-chan net.ThreadRecord, error)
+
+// NewConnector creates bidirectional connection between an app and a thread.
+func NewConnector(app App, net Net, tinfo thread.Info, conn Connection) (*Connector, error) {
+	if !tinfo.Key.CanRead() {
+		log.Fatalf("read key not found for thread %s", tinfo.ID)
+	}
+	lg := tinfo.GetOwnLog()
+	if lg == nil {
+		return nil, fmt.Errorf("own log for thread %s does not exist", tinfo.ID)
+	}
+	a := &Connector{
+		app:       app,
+		net:       net,
+		threadID:  tinfo.ID,
+		threadKey: tinfo.Key,
+		logID:     lg.ID,
+		closeChan: make(chan struct{}),
+	}
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go a.threadToApp(conn, &wg)
+	go a.appToThread(&wg)
+	wg.Wait()
+	a.goRoutines.Add(2)
+	return a, nil
+}
+
+// Close stops the bidirectional connection between the app and thread.
+func (c *Connector) Close() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.closed {
+		return nil
+	}
+	c.closed = true
+	close(c.closeChan)
+	c.goRoutines.Wait()
+	return nil
+}
+
+// ThreadID returns the underlying thread's ID.
+func (c *Connector) ThreadID() thread.ID {
+	return c.threadID
+}
+
+func (c *Connector) threadToApp(con Connection, wg *sync.WaitGroup) {
+	defer c.goRoutines.Done()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	filter := map[thread.ID]struct{}{c.threadID: {}}
+	sub, err := con(ctx, filter)
+	if err != nil {
+		log.Fatalf("error getting thread subscription: %v", err)
+	}
+	wg.Done()
+	for {
+		select {
+		case <-c.closeChan:
+			log.Debugf("closing thread-to-app flow on thread %s", c.threadID)
+			return
+		case rec, ok := <-sub:
+			if !ok {
+				log.Debug("notification channel closed, not listening to external changes anymore")
+				return
+			}
+			if err = c.app.HandleNetRecord(rec, c.threadKey, c.logID, fetchEventTimeout); err != nil {
```

Comment:
This method now takes a thread key, which is now baked into the connection. There's no longer a need to fetch the thread here.